includes/notes/class-wc-admin-notes-order-milestones.php

```diff
@@ -0,0 +1,262 @@
+<?php
+/**
+ * WooCommerce Admin (Dashboard) Order Milestones Note Provider.
+ *
+ * Adds a note to the merchant's inbox when certain order milestones are reached.
+ *
+ * @package WooCommerce Admin
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * WC_Admin_Notes_Order_Milestones
+ */
+class WC_Admin_Notes_Order_Milestones {
+	/**
+	 * Name of the "first order" note.
+	 */
+	const FIRST_ORDER_NOTE_NAME = 'wc-admin-first-order';
+
+	/**
+	 * Name of the "ten orders" note.
+	 */
+	const TEN_ORDERS_NOTE_NAME = 'wc-admin-ten-orders';
+
+	/**
+	 * Name of the "other milestones" note.
+	 */
+	const ORDERS_MILESTONE_NOTE_NAME = 'wc-admin-orders-milestone';
+
+	/**
+	 * Option key name to store last order milestone.
+	 */
+	const LAST_ORDER_MILESTONE_OPTION_KEY = 'wc_admin_last_orders_milestone';
+
+	/**
+	 * Hook to process order milestones.
+	 */
+	const PROCESS_ORDERS_MILESTONE_HOOK = 'wc_admin_process_orders_milestone';
+
+	/**
+	 * Allowed order statuses for calculating milestones.
+	 *
+	 * @var array
+	 */
+	protected $allowed_statuses = array(
+		'pending',
+		'processing',
+		'completed',
+	);
+
+	/**
+	 * Orders count cache.
+	 *
+	 * @var int
+	 */
+	protected $orders_count = null;
+
+	/**
+	 * Further order milestone thresholds.
+	 *
+	 * @var array
+	 */
+	protected $milestones = array(
+		100,
+		250,
+		500,
+		1000,
+		5000,
+		10000,
+		500000,
+		1000000,
+	);
+
+	/**
+	 * Delay hook attachment until after the WC post types have been registered.
+	 *
+	 * This is required for retrieving the order count.
+	 */
+	public function __construct() {
+		/**
+		 * Filter Order statuses that will count towards milestones.
+		 *
+		 * @since 3.5.0
+		 *
+		 * @param array $allowed_statuses Order statuses that will count towards milestones.
+		 */
+		$this->allowed_statuses = apply_filters( 'woocommerce_admin_order_milestone_statuses', $this->allowed_statuses );
+
+		add_action( 'woocommerce_after_register_post_type', array( $this, 'init' ) );
+	}
+
+	/**
+	 * Hook everything up.
+	 */
+	public function init() {
+		if ( ! wp_next_scheduled( self::PROCESS_ORDERS_MILESTONE_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::PROCESS_ORDERS_MILESTONE_HOOK );
```

Comment:
Dummy question here, but shouldn't we add a [`wp_clear_scheduled_hook()`](https://codex.wordpress.org/Function_Reference/wp_clear_scheduled_hook) somewhere to clear this event when the plugin is deactivated?

Reply:
Good catch - added in 37db77b.